File: src/Night.Ms.SshServer/Providers/Finance/Sparkline.cs
namespace Night.Ms.SshServer.Providers.Finance;

// Tiny block-glyph chart utilities used by FinanceScreen + FinanceDetailScreen.
//
// Sparkline.Render is the one-line inline chart: each output character represents one bucket
// from the input series, scaled into the 8-step block alphabet.
//
// BigChart.Render is the multi-row detail chart: each output column is one bucket painted
// using full + partial top-blocks so the chart reads as a continuous line.
public static class Sparkline
{
    // ▁ U+2581 .. █ U+2588 — eight ascending lower-block glyphs.
    private const string Blocks = "▁▂▃▄▅▆▇█";

    public static string Render(IReadOnlyList<double>? series, int width)
    {
        if (series is null || series.Count == 0 || width <= 0) return string.Empty;

        var buckets = Bucketize(series, width);
        if (buckets.Length == 0) return string.Empty;

        var min = double.PositiveInfinity;
        var max = double.NegativeInfinity;
        foreach (var v in buckets)
        {
            if (double.IsNaN(v) || double.IsInfinity(v)) continue;
            if (v < min) min = v;
            if (v > max) max = v;
        }
        if (double.IsInfinity(min)) return string.Empty;

        var range = max - min;
        var result = new char[buckets.Length];
        for (var i = 0; i < buckets.Length; i++)
        {
            var v = buckets[i];
            if (double.IsNaN(v) || double.IsInfinity(v))
            {
                result[i] = ' ';
                continue;
            }
            // Constant series → all bars at the lowest level so the user sees a flat line
            // rather than nothing.
            var t = range > 0 ? (v - min) / range : 0d;
            var idx = (int)Math.Round(t * (Blocks.Length - 1));
            if (idx < 0) idx = 0;
            else if (idx >= Blocks.Length) idx = Blocks.Length - 1;
            result[i] = Blocks[idx];
        }
        return new string(result);
    }

    // Bucketing strategy: when the series is longer than width, average values per bucket;
    // when it's shorter, repeat values so the chart fills the full width (small series still
    // look like a chart, not a stub). NaN/Infinity inputs are dropped from the average.
    private static double[] Bucketize(IReadOnlyList<double> series, int width)
    {
        var clean = new List<double>(series.Count);
        foreach (var v in series)
            if (!double.IsNaN(v) && !double.IsInfinity(v)) clean.Add(v);
        if (clean.Count == 0) return [];

        if (clean.Count <= width)
        {
            var outArr = new double[clean.Count];
            for (var i = 0; i < clean.Count; i++) outArr[i] = clean[i];
            return outArr;
        }

        var buckets = new double[width];
        var step = (double)clean.Count / width;
        for (var i = 0; i < width; i++)
        {
            var from = (int)Math.Floor(i * step);
            var to = (int)Math.Floor((i + 1) * step);
            if (to <= from) to = from + 1;
            if (to > clean.Count) to = clean.Count;
            var sum = 0d;
            var n = 0;
            for (var j = from; j < to; j++) { sum += clean[j]; n++; }
            buckets[i] = n > 0 ? sum / n : double.NaN;
        }
        return buckets;
    }
}

// Multi-row chart used by FinanceDetailScreen. Output is a list of strings, top row first.
// Each column shows the bucketed value as filled rows from the bottom, with the topmost row
// using a partial block glyph for sub-row resolution.
public static class BigChart
{
    private const string Blocks = "▁▂▃▄▅▆▇█";

    public static IReadOnlyList<string> Render(IReadOnlyList<double>? series, int width, int height)
    {
        if (series is null || series.Count == 0 || width <= 0 || height <= 0)
            return Enumerable.Repeat(string.Empty, Math.Max(0, height)).ToList();

        // Reuse Sparkline's bucketing — same min/max scan logic up here so the two charts
        // never disagree on which buckets are missing.
        var buckets = BucketAverages(series, width);
        if (buckets.Length == 0)
            return Enumerable.Repeat(new string(' ', width), height).ToList();

        var min = double.PositiveInfinity;
        var max = double.NegativeInfinity;
        foreach (var v in buckets)
        {
            if (double.IsNaN(v)) continue;
            if (v < min) min = v;
            if (v > max) max = v;
        }
        if (double.IsInfinity(min))
            return Enumerable.Repeat(new string(' ', width), height).ToList();

        var range = max - min;
        var rows = new char[height][];
        for (var r = 0; r < height; r++) { rows[r] = new char[width]; Array.Fill(rows[r], ' '); }

        for (var c = 0; c < buckets.Length; c++)
        {
            var v = buckets[c];
            if (double.IsNaN(v)) continue;
            var t = range > 0 ? (v - min) / range : 0.5;
            // Total fill in 1/8-row units, capped at the chart's full height.
            var subRows = (int)Math.Round(t * height * 8);
            if (subRows < 1) subRows = 1;
            if (subRows > height * 8) subRows = height * 8;

            var fullRows = subRows / 8;
            var partial = subRows % 8;

            for (var r = 0; r < fullRows; r++)
                rows[height - 1 - r][c] = '█';
            if (partial > 0 && fullRows < height)
                rows[height - 1 - fullRows][c] = Blocks[partial - 1];
        }

        var result = new string[height];
        for (var r = 0; r < height; r++) result[r] = new string(rows[r]);
        return result;
    }

    private static double[] BucketAverages(IReadOnlyList<double> series, int width)
    {
        var clean = new List<double>(series.Count);
        foreach (var v in series)
            if (!double.IsNaN(v) && !double.IsInfinity(v)) clean.Add(v);
        if (clean.Count == 0) return [];

        if (clean.Count <= width)
        {
            // Stretch a short series across `width` columns so the chart fills the box.
            var stretched = new double[width];
            for (var i = 0; i < width; i++)
            {
                var srcIdx = (int)Math.Floor((double)i * clean.Count / width);
                if (srcIdx >= clean.Count) srcIdx = clean.Count - 1;
                stretched[i] = clean[srcIdx];
            }
            return stretched;
        }

        var buckets = new double[width];
        var step = (double)clean.Count / width;
        for (var i = 0; i < width; i++)
        {
            var from = (int)Math.Floor(i * step);
            var to = (int)Math.Floor((i + 1) * step);
            if (to <= from) to = from + 1;
            if (to > clean.Count) to = clean.Count;
            var sum = 0d;
            var n = 0;
            for (var j = from; j < to; j++) { sum += clean[j]; n++; }
            buckets[i] = n > 0 ? sum / n : double.NaN;
        }
        return buckets;
    }
}

File: internal/tui/components/sparkline.go
// Package components — block-glyph chart utilities used by the finance screens.
//
// Sparkline is the one-line inline chart (each output character represents one
// bucket from the input series, scaled into the 8-step block alphabet).
//
// BigChart is the multi-row detail chart (each column shows the bucketed value
// as filled rows from the bottom, with the topmost row using a partial-block
// glyph for sub-row resolution).
//
// Ported from ssh.night.ms/.../Providers/Finance/Sparkline.cs.
package components

import (
	"math"
	"strings"
)

// blocks is U+2581 .. U+2588 — eight ascending lower-block glyphs.
var blocks = []rune{'▁', '▂', '▃', '▄', '▅', '▆', '▇', '█'}

// Sparkline renders a series of numbers as a single line of block glyphs.
// Returns empty string when series is empty / width <= 0 / all values are
// NaN+Inf. A constant series renders as a flat row of the lowest glyph.
func Sparkline(series []float64, width int) string {
	if len(series) == 0 || width <= 0 {
		return ""
	}
	buckets := bucketize(series, width, false)
	if len(buckets) == 0 {
		return ""
	}
	min, max, ok := minMaxFinite(buckets)
	if !ok {
		return ""
	}
	rang := max - min
	out := make([]rune, len(buckets))
	for i, v := range buckets {
		if math.IsNaN(v) || math.IsInf(v, 0) {
			out[i] = ' '
			continue
		}
		t := 0.0
		if rang > 0 {
			t = (v - min) / rang
		}
		idx := int(math.Round(t * float64(len(blocks)-1)))
		if idx < 0 {
			idx = 0
		} else if idx >= len(blocks) {
			idx = len(blocks) - 1
		}
		out[i] = blocks[idx]
	}
	return string(out)
}

// BigChart renders a series as a height-row, width-column block chart. Returns
// `height` strings (top row first). Short series are stretched to fill width
// so a small chart still reads as a chart, not a stub.
func BigChart(series []float64, width, height int) []string {
	if len(series) == 0 || width <= 0 || height <= 0 {
		out := make([]string, 0, height)
		for i := 0; i < height; i++ {
			out = append(out, "")
		}
		return out
	}
	buckets := bucketize(series, width, true)
	blank := strings.Repeat(" ", width)
	if len(buckets) == 0 {
		out := make([]string, height)
		for i := range out {
			out[i] = blank
		}
		return out
	}
	min, max, ok := minMaxFinite(buckets)
	if !ok {
		out := make([]string, height)
		for i := range out {
			out[i] = blank
		}
		return out
	}
	rang := max - min
	rows := make([][]rune, height)
	for r := range rows {
		rows[r] = []rune(blank)
	}
	for c, v := range buckets {
		if math.IsNaN(v) || math.IsInf(v, 0) {
			continue
		}
		t := 0.5
		if rang > 0 {
			t = (v - min) / rang
		}
		subRows := int(math.Round(t * float64(height) * 8))
		if subRows < 1 {
			subRows = 1
		}
		if subRows > height*8 {
			subRows = height * 8
		}
		fullRows := subRows / 8
		partial := subRows % 8
		for r := 0; r < fullRows; r++ {
			rows[height-1-r][c] = '█'
		}
		if partial > 0 && fullRows < height {
			rows[height-1-fullRows][c] = blocks[partial-1]
		}
	}
	out := make([]string, height)
	for r := range out {
		out[r] = string(rows[r])
	}
	return out
}

// bucketize partitions a series into `width` buckets averaging values per
// bucket. NaN/Inf inputs are dropped before bucketing. When the series is
// shorter than width: Sparkline (stretch=false) returns the cleaned values
// unchanged; BigChart (stretch=true) repeats values so columns fill the box.
func bucketize(series []float64, width int, stretch bool) []float64 {
	clean := make([]float64, 0, len(series))
	for _, v := range series {
		if !math.IsNaN(v) && !math.IsInf(v, 0) {
			clean = append(clean, v)
		}
	}
	if len(clean) == 0 {
		return nil
	}
	if len(clean) <= width {
		if !stretch {
			return clean
		}
		out := make([]float64, width)
		for i := 0; i < width; i++ {
			src := int(math.Floor(float64(i) * float64(len(clean)) / float64(width)))
			if src >= len(clean) {
				src = len(clean) - 1
			}
			out[i] = clean[src]
		}
		return out
	}
	out := make([]float64, width)
	step := float64(len(clean)) / float64(width)
	for i := 0; i < width; i++ {
		from := int(math.Floor(float64(i) * step))
		to := int(math.Floor(float64(i+1) * step))
		if to <= from {
			to = from + 1
		}
		if to > len(clean) {
			to = len(clean)
		}
		sum, n := 0.0, 0
		for j := from; j < to; j++ {
			sum += clean[j]
			n++
		}
		if n > 0 {
			out[i] = sum / float64(n)
		} else {
			out[i] = math.NaN()
		}
	}
	return out
}

func minMaxFinite(vs []float64) (min, max float64, ok bool) {
	min, max = math.Inf(1), math.Inf(-1)
	for _, v := range vs {
		if math.IsNaN(v) || math.IsInf(v, 0) {
			continue
		}
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
		ok = true
	}
	return
}
